### PR TITLE
fix!: deal rewards

### PR DESCRIPTION
**Pull request by elshan-eth**
- mv capacity const to core (affect CLI and NOX)
- fix by audit-report
- fix vesting and reward pool calculation

### CODE REVIEW

File: src/deal/Deal.sol [hunk 1]
```diff
@@ -374,94 +200,178 @@ contract Deal is MulticallUpgradeable, WorkerManager, IDeal {
...
+        }
+
+        uint256 currentWorkerCount = getWorkerCount();
         uint256 pricePerWorkerEpoch_ = pricePerWorkerEpoch();
+        snapshot =
+            DealSnapshot.init(dealStorage, currentEpoch, commitEpoch, isEnded, pricePerWorkerEpoch_, currentWorkerCount);
 
-        uint256 workerCount = getWorkerCount();
+        uint256 lastCommitedEpoch = dealStorage.lastCommitedEpoch;
+        if (commitEpoch <= lastCommitedEpoch) {
```
Comment (ValeryAntopol) [inline src/deal/Deal.sol:288]: why is this check is not at the start of the method?

File: src/deal/Deal.sol [hunk 2]
```diff
@@ -374,94 +200,178 @@ contract Deal is MulticallUpgradeable, WorkerManager, IDeal {
...
+                snapshot.setLockedBalance(snapshot.getLockedBalance() + amount); // record locked balance for rewards
+                snapshot.setGapsEpochCount(snapshot.getGapsEpochCount() + (commitEpoch - maxPaidEpoch)); // write gaps
+            }
+        } else {
+            // commitEpoch <= maxPaidEpoch, this means that we we need to record only active Epochs and write off balances. We don't have a gaps.
+            uint256 amount = (commitEpoch - lastCommitedEpoch) * pricePerWorkerEpoch_ * currentWorkerCount;
 
-        emit DealEnded(currentEpoch);
+            snapshot.setTotalBalance(snapshot.getTotalBalance() - amount); // write off balance
+            snapshot.setLockedBalance(snapshot.getLockedBalance() + amount); // record locked balance for rewards
```
Comment (ValeryAntopol) [inline src/deal/Deal.sol:312]: maybe decouple this part into a function? this code is doubled in the other branch of if

Comment (elshan-eth) [reply src/deal/Deal.sol:312]: I think we’ll do it later

File: src/deal/Deal.sol (hunk 1 above)
Comment (elshan-eth) [reply src/deal/Deal.sol:288]: Because before we init snapshot

File: integration-tests/src/commitment.test.ts [hunk 3]
```diff
@@ -235,16 +235,17 @@ describe("Capacity commitment", () => {
...
     const HACKY_WITHDRAW_EPOCHS = withdrawEpochs * 2n;
     await skipEpoch(provider, epochDuration, HACKY_WITHDRAW_EPOCHS);
 
+    // TODO: Add error class
     await expect(
       capacityContract
         .finishCommitment(commitmentId)
         .then((tx) => tx.wait(DEFAULT_CONFIRMATIONS)),
-    ).rejects.toThrow("For finish commitment all units should be exited");
+    ).rejects.toThrow("(unknown custom error)");
```
Comment (folex) [inline integration-tests/src/commitment.test.ts:248]: 😭

Comment (akim-bow) [reply integration-tests/src/commitment.test.ts:248]: Я уже готовлю фикс для этого в след пулике. Это чтоб тесты прошли

**Pull request merged**